
<!-- README.md is generated from README.Rmd. Please edit that file -->

# P2C2M2

<!-- badges: start -->
<!-- badges: end -->

P2C2M2 provides functions to read default output from BEAST 2 (Bouckaert
et al. 2019) and StarBEAST2 (Ogilvie et al. 2017) and conduct posterior
predictive checks of coalescent models (Reid et al. 2014) based on the
functions originally provided in the P2C2M package (Gruenstaeudl et
al. 2016) to parse tree and xml files from BEAST and \*BEAST. It also
implements estimation of type-I (false-positive) error rates using
pseudo-observed datasets (“pods”) sampled from the posterior predictive
distribution.

## Installation

You can install the development version of P2C2M2 from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("arleyc/P2C2M2")
```

## Input files

In order to generate the tree files annotated with the required branch
rate estimates, the XML file has to be modified by hand to include the
following code before running BEAST 2:

\#For strict clocks:

Insert the following:<br> <em>substitutions=“true”
branchratemodel=“@StrictClock.c:Locus_Name”</em><br> after
<em>id=“TreeWithMetaDataLogger.t:Locus_Name”</em> and before
<em>spec=“beast.base.evolution.TreeWithMetaDataLogger”</em>

\#For UCLN or RLC clocks:

Insert the following:<br> <em>substitutions=“true”</em><br> after
<em>id=“TreeWithMetaDataLogger.t:Locus_Name”</em> and before
<em>spec=“beast.base.evolution.TreeWithMetaDataLogger”</em>

## Example

Basic example showing how to run the full pipeline with a single
function using the provided dataset to test the multispecies coalescent
model and to estimate false positive error rates per locus and across
loci:

``` r
library(P2C2M2)

# The absolute path to the input directory is set
inPath <- system.file("extdata", "thomomys/", package="P2C2M2")

# The name of the xml-file generated by BEAUTi and located in 
# "inPath" is set
inFile <- "thomomys.xml"

# Posterior predictive simulations with a setting of 2 simulation 
# replicates are preformed
suppressWarnings(thomomys <- p2c2m2.complete(inPath, inFile, num.reps=2, error.rate = TRUE))
#> 
#> bindings, loci, and alignments read
#> 
#> locus 26 is done
#> 
#> locus 29 is done
#> 
#> locus 47 is done
#> 
#> locus 53 is done
#> 
#> locus 59 is done
#> 
#> locus 64 is done
#> 
#> locus 72 is done
#> 
#> gene trees done
#> 
#> species trees done
#> 
#> Calculating metrics for 3 trees of the posterior distribution
#> 
#> Calculating metrics for the trees of the post. predictive distribution
#> 
#> Calculating differences btw. the posterior and the post. predictive distribution
#> 
#> Calculating differences btw. 1000 pods and the post. predictive distribution to estimate error rates

thomomys$results$alpha0.01
#> $perGene
#>    LCWT[2]                NDC[2]               
#> 26 "-16.8 (±37.28) n.s."  "-6.33 (±11.15) n.s."
#> 29 "-11.94 (±16.13) n.s." "-0.67 (±4.51) n.s." 
#> 47 "18.92 (±16.41) *"     "2 (±7) n.s."        
#> 53 "-17.05 (±8.08) *"     "2 (±1.73) *"        
#> 59 "-0.72 (±33.96) n.s."  "-5.67 (±5.03) *"    
#> 64 "-33.34 (±4.29) *"     "13.33 (±4.16) *"    
#> 72 "20.16 (±20.77) n.s."  "-8.67 (±3.06) *"    
#> 
#> $acrGenes
#>        LCWT[2]                NDC[2]               
#> Sum    "-29.62 (±69.31) n.s." "-6.14 (±13.15) n.s."
#> Mean   "-4.23 (±9.9) n.s."    "-0.88 (±1.88) n.s." 
#> Median "-2.17 (±25.62) n.s."  "-0.14 (±6.36) n.s." 
#> Mode   "5.33 (±27.19) n.s."   "-2.71 (±12.94) n.s."
#> CV     "0.47 (±4.52) n.s."    "-4.74 (±7.31) n.s." 
#> 
#> $perGene.error
#>    LCWT[2] NDC[2]
#> 26   0.649  0.657
#> 29   0.669  0.643
#> 47   0.649  0.643
#> 53   0.649  1.000
#> 59   0.649  0.700
#> 64   0.669  1.000
#> 72   0.669  1.000
#> 
#> $acrGenes.error
#>        LCWT[2] NDC[2]
#> Sum      0.669  0.700
#> Mean     0.669  0.700
#> Median   0.000  0.343
#> Mode     0.000  0.000
#> CV       0.669  0.700

thomomys$results$legend
#> [1] "Differences between the posterior and the posterior predictive distributions per locus and across loci. Each cell contains the following information in said order: mean, standard deviation, significance level. Error rates (if estimated with option error.rate=TRUE) are based on differences between the pods and the posterior predictive distributions. Codes in square brackets indicate the number of tails. Alpha values are automatically adjusted for the number of tails."
```

## References

Bouckaert, R., Vaughan, T.G., Barido-Sottani, J., Duchêne, S., Fourment,
M., Gavryushkina, A., et al. (2019) BEAST 2.5: An advanced software
platform for Bayesian evolutionary analysis. PLoS Computational Biology,
15(4), e1006650.

Ogilvie, H.A., Bouckaert, R.R., Drummond, A.J. (2017) StarBEAST2 brings
faster species tree inference and accurate estimates of substitution
rates. Molecular Biology and Evolution, 34(8), 2101-–2114.

Gruenstaeudl, M., Reid, N.M., Wheeler, G.R. and Carstens, B.C. (2016)
Posterior predictive checks of coalescent nodels: P2C2M, an R package.
Molecular Ecology Resources, 6, 193–205.

Reid, N.M., Brown, J.M., Satler, J.D., Pelletier, T.A., McVay, J.D.,
Hird, S.M. and Carstens, B.C. (2014) Poor fit to the multi-species
coalescent model is widely detectable in empirical data. Systematic
Biology, 63, 322–333.

## Citation

Villamil, J., Morando, M., Avila, L.J., Lanna, F.M., Fonseca, E.M.,
Sites, Jr., J.W., Camargo, A. Revisiting the problem of Multispecies
Coalescent Model fit with an example from the first complete molecular
phylogeny of the <em>Liolaemus wiegmannii</em> species group (Squamata:
Liolaemidae), in review.
